File: run_digital_twin_framework.m
%% ============================================================
% FAILURE SIMULATION FRAMEWORK FOR PREDICTIVE DIGITAL TWINS
% Main Execution Script (Corrected - Full Pipeline)
% ============================================================

clc
clear
close all

fprintf('Starting Failure Simulation Framework...\n')

num_machines = 4;
t  = 1:12000;
dt = 1;

%% --- Step 1: Run Multi-Asset Digital Twin Simulation ---
machines = digital_twin_simulatorx(num_machines, t, dt);

%% --- Step 2: Fleet Failure Risk Ranking ---
fleet_failure_ranking(machines)

fprintf('\nSimulation Complete.\n')

%% ============================================================
% PROBABILISTIC RUL ANALYSIS (Machine 1)
% ============================================================

machine = machines(1);
num_sim = 100;
k = 1e-4;
a = 1.5;
b = 1.2;
c = 1.1;
d = 1.3;

%% --- Step 3: Monte Carlo RUL Simulation ---
failure_times = monte_carlo_rul(machine, t, dt, k, a, b, c, d, num_sim);

%% Compute statistics
mean_tf  = mean(failure_times);
std_tf   = std(failure_times);

%% --- Step 4: Confidence Interval ---
lower_95 = mean_tf - 2*std_tf;
upper_95 = mean_tf + 2*std_tf;

%% Plot failure time distribution
figure
histogram(failure_times, 20)
xlabel('Failure Time')
ylabel('Frequency')
title('Failure Time Distribution with Confidence Interval')
grid on
hold on
xline(mean_tf,   'r',   'LineWidth', 2, 'Label', 'Mean')
xline(lower_95,  'g--', 'LineWidth', 2, 'Label', 'Lower 95%')
xline(upper_95,  'g--', 'LineWidth', 2, 'Label', 'Upper 95%')

fprintf("\n=== Probabilistic RUL Results (Machine 1) ===\n")
fprintf("Mean Failure Time: %.2f\n",  mean_tf)
fprintf("Std Dev: %.2f\n",            std_tf)
fprintf("95%% Confidence Interval: [%.2f, %.2f]\n", lower_95, upper_95)

threshold_time = round(mean_tf);
prob_failure   = sum(failure_times <= threshold_time) / length(failure_times);
fprintf("Probability of failure before %d: %.2f\n", threshold_time, prob_failure)

%% --- Step 5: CDF (Failure Probability Curve) ---
sorted_times = sort(failure_times);
cdf = (1:length(sorted_times)) / length(sorted_times);

figure
plot(sorted_times, cdf, 'LineWidth', 2)
xlabel('Time')
ylabel('Probability of Failure')
title('Failure Probability Curve (CDF)')
grid on

File: run_digital_twin_framework.py
# ============================================================
# FAILURE SIMULATION FRAMEWORK FOR PREDICTIVE DIGITAL TWINS
# Main Execution Script
# ============================================================

import numpy as np
import matplotlib.pyplot as plt

from digital_twin_simulatorx import digital_twin_simulatorx
from fleet_failure_ranking   import fleet_failure_ranking
from monte_carlo_rul         import monte_carlo_rul

def main():

    print("Starting Failure Simulation Framework...")

    num_machines = 4
    t  = np.arange(1, 12001)
    dt = 1

    machines = digital_twin_simulatorx(num_machines, t, dt)
    fleet_failure_ranking(machines)

    print("\nSimulation Complete.")

    machine = machines[0]
    num_sim = 100
    k = 1e-4; a = 1.5; b = 1.2; c = 1.1; d = 1.3

    failure_times = monte_carlo_rul(machine, t, dt, k, a, b, c, d, num_sim)

    mean_tf  = np.mean(failure_times)
    std_tf   = np.std(failure_times)
    lower_95 = mean_tf - 2*std_tf
    upper_95 = mean_tf + 2*std_tf

    plt.figure()
    plt.hist(failure_times, bins=20)
    plt.xlabel('Failure Time')
    plt.ylabel('Frequency')
    plt.title('Failure Time Distribution with Confidence Interval')
    plt.grid(True)
    plt.axvline(mean_tf,  color='r', linewidth=2, linestyle='-',  label='Mean')
    plt.axvline(lower_95, color='g', linewidth=2, linestyle='--', label='Lower 95%')
    plt.axvline(upper_95, color='g', linewidth=2, linestyle='--', label='Upper 95%')
    plt.legend()

    print(f"\n=== Probabilistic RUL Results (Machine 1) ===")
    print(f"Mean Failure Time: {mean_tf:.2f}")
    print(f"Std Dev: {std_tf:.2f}")
    print(f"95% Confidence Interval: [{lower_95:.2f}, {upper_95:.2f}]")

    threshold_time = round(mean_tf)
    prob_failure   = np.sum(failure_times <= threshold_time) / len(failure_times)
    print(f"Probability of failure before {threshold_time}: {prob_failure:.2f}")

    sorted_times = np.sort(failure_times)
    cdf = np.arange(1, len(sorted_times) + 1) / len(sorted_times)

    plt.figure()
    plt.plot(sorted_times, cdf, linewidth=2)
    plt.xlabel('Time')
    plt.ylabel('Probability of Failure')
    plt.title('Failure Probability Curve (CDF)')
    plt.grid(True)

    plt.show()


if __name__ == "__main__":
    main()